hal/api/EventAdapter.h

```diff
@@ -0,0 +1,316 @@
+     *  @param func     Static function to attach
+     *  @param arg      Pointer argument to function
+     */
+    template<typename T>
+    EventAdapter(void (*func)(const T*), const T *arg) {
+        construct_callback(callback(func, arg));
+    }
+
+    /** Create an EventAdapter with a static function and bound pointer callback
+     *  @param func     Static function to attach
+     *  @param arg      Pointer argument to function
+     */
+    template<typename T>
+    EventAdapter(void (*func)(volatile T*), volatile T *arg) {
+        construct_callback(callback(func, arg));
+    }
+
+    /** Create an EventAdapter with a static function and bound pointer callback
+     *  @param func     Static function to attach
+     *  @param arg      Pointer argument to function
+     */
+    template<typename T>
+    EventAdapter(void (*func)(const volatile T*), const volatile T *arg) {
+        construct_callback(callback(func, arg));
+    }
+
+#if MBED_CONF_MBED_EVENTS_PRESENT
+    /** Create an EventAdapter that holds an Event
+     *  @param evt The event
+     */
+    EventAdapter(const adapter_event_t &evt) {
+        construct_event(evt);
+    }
+#endif
+
+    EventAdapter(const EventAdapter &adapter) {
+        _attach(adapter);
+    }
+
+    /** Test if a function is attached to this adapter
+     */
+    operator bool() const {
+        if (_is_cb) {
+            return get_callback()->operator bool();
+        } else {
+            return true;
+        }
+    }
+
+    EventAdapter& operator =(const EventAdapter &rhs) {
+        if (this != &rhs) {
+            destroy();
+            _attach(rhs);
+        }
+        return *this;
+    }
+
+    /** Attach this adapter to the callback or event in a different adapter
+     *  @param adapter The EventAdapter to attach.
+     */
+    void attach(const EventAdapter &adapter) {
+        destroy();
+        _attach(adapter);
+    }
+
+    /** Call the callback or event in this adapter
+     */
+    void call() {
+        if (_is_cb) {
+            get_callback()->call();
+        } else {
+#if MBED_CONF_MBED_EVENTS_PRESENT
+            get_event()->call();
+#endif
+        }
+    }
+
+    /** Call the callback or event in this adapter
+     */
+    void operator ()() {
+        call();
+    }
+
+    ~EventAdapter() {
+        destroy();
+    }
+
+private:
+    // No unions for non-POD data types in C++03. Need to use placement new to simulate the
+    // union between a Callback<void()> and an Event<void()>
+    union {
+        // Ensure alignment by declaring the storage uint32_t instead of uint8_t
+        uint32_t _cb_data[(sizeof(adapter_callback_t) + sizeof(uint32_t) - 1) / sizeof(uint32_t)];
+#if MBED_CONF_MBED_EVENTS_PRESENT
+        uint32_t _evt_data[(sizeof(adapter_event_t) + sizeof(uint32_t) - 1) / sizeof(uint32_t)];
+#endif
+    };
+    bool  _is_cb;
+
+    void destroy() {
+        if (_is_cb) {
+            get_callback()->~adapter_callback_t();
```

Comment:
Not sure what you mean by that. That's a manual destructor call, needed because of a previous call to the placement `new` operator. I don't think that overloading `->` helps you in this situation.